### ISSUE TITLE
Strange rename issue with MinIO + CloudBerry combo

### ISSUE THREAD

**Issue by mhjartstrom** [minio/minio#8234]
## Expected Behavior

I'm running MinIO as a stand-alone executable and recently I get huge amounts of error messages while using CloudBerry, saying the rename operation fails because it can't move the file to another disk drive. Naturally this should not happen.

## Current Behavior

I get huge amounts of this kind of error message:
`API: PutObject(bucket=projekt, object=CBB_SERVER/CBB_Configuration/9c63d888-fad9-4bcf-8069-d41b82cbf533.cbb$/20190810140312/9c63d888-fad9-4bcf-8069-d41b82cbf533.cbb)`
`Time: 13:29:45 CEST 09/03/2019`
`DeploymentID: 35af7469-a8ab-4899-9a25-58ae7b15a284`
`RequestID: 15C0EA353C8E2420`
`RemoteHost: <IP>`
`Host: <IP>:9001`
`UserAgent: CloudBerryLab.Base.HttpUtil.Client 5.9.2 (http://www.cloudberrylab.com/)`
`Error: rename F:\MinIO\.minio.sys/tmp/cb0e8a6f-e0ff-445d-8297-bf0d49501f14/6d594026-7948-4f2d-a44a-8ea3721398ad F:\MinIO\projekt/CBB_SERVER/CBB_Configuration/9c63d888-fad9-4bcf-8069-d41b82cbf533.cbb$/20190810140312/9c63d888-fad9-4bcf-8069-d41b82cbf533.cbb: Det går inte att flytta filen till en annan diskenhet.`
`       5: cmd\fs-v1-helpers.go:413:cmd.fsRenameFile()`
`       4: cmd\fs-v1.go:931:cmd.(*FSObjects).putObject()`
`       3: cmd\fs-v1.go:818:cmd.(*FSObjects).PutObject()`
`       2: cmd\object-handlers.go:1251:cmd.objectAPIHandlers.PutObjectHandler()`
`       1: net\http\server.go:1995:http.HandlerFunc.ServeHTTP()`

I execute minio.exe similar to:
`C:\Admin>minio server --address :9001 --certs-dir "C:\Admin\.minio\certs" --config-dir "C:\Admin\.minio" "F:\MinIO\"`

## Steps to Reproduce (for bugs)

1. Run MinIO as a stand-alone executable using a command line similar to the one described.
2. Try to make a CloudBerry Backup backup plan using the MinIO server as the target
3. The CloudBerry log files will say "The remote server returned an error: (500) Internal Server Error."
4. Watch the rename error message from the MinIO console output.

## Context

All my CloudBerry backups are failing.

## Regression

It has been working previously, but I have a feeling the bug was revealed by upgrading to one of the latest CloudBerry releases. Last successful backup was performed on 2019-08-12.

## Your Environment

* Version used (`minio version`): 2019-09-11T19:53:16Z
* Environment name and version (e.g. nginx 1.9.1): Windows 7 SP1
* Server type and version:
* Operating System and version (`uname -a`): Windows 7 SP1
* Link to your project:

**Comment by harshavardhana** [comment]
> `UserAgent: CloudBerryLab.Base.HttpUtil.Client 5.9.2 (http://www.cloudberrylab.com/)`
> `Error: rename F:\MinIO\.minio.sys/tmp/cb0e8a6f-e0ff-445d-8297-bf0d49501f14/6d594026-7948-4f2d-a44a-8ea3721398ad F:\MinIO\projekt/CBB_SERVER/CBB_Configuration/9c63d888-fad9-4bcf-8069-d41b82cbf533.cbb$/20190810140312/9c63d888-fad9-4bcf-8069-d41b82cbf533.cbb: Det går inte att flytta filen till en annan diskenhet.`

Do you have english translation of the error message?

**Comment by mhjartstrom** [comment]
Yes, as I said under "Expected Behavior" the Swedish phrase "Det går inte att flytta filen till en annan diskenhet." means "Unable to move file to another disk drive.".

**Comment by kannappanr** [comment]
@mhjartstrom The command should be

```
minio server --address :9001 --certs-dir "C:\Admin\.minio\certs" "F:\MinIO\"
```

**Comment by kannappanr** [comment]
@mhjartstrom From the error it looks like that `F:\MinIO\.minio.sys` and `F:\MinIO\projekt` point to different drives. In the case of `FS` MinIO does not support renaming files to another drive.

**Comment by mhjartstrom** [comment]
Removing the `--config-dir "C:\Admin\.minio"` argument has no effect whatsoever on this issue, apart from making MinIO create a `.minio` directory under the user home directory for no good reason. I have my config directory under `C:\Admin\.minio` for all instances, not under the user home directory.

On a Windows system `F:` is `F:`, there is no different drive here... `F:\MinIO\.minio.sys` and `F:\MinIO\projekt` does not point to different drives, as they are both on F: and in no way are they any kind of soft or hard link to something else. They are simple NTFS directories on the same drive.

The error description does use two different path notations, i.e. `\` and `/` in the same path `F:\MinIO\projekt/CBB_SERVER/CBB_Configuration/...`. Is that handled properly in the underlying system?

**Comment by harshavardhana** [comment]
Windows handles both paths and they are automatically translated by Go os package.

**Comment by mhjartstrom** [comment]
Then I don't understand what the error message means. It's obviously trying to rename/move a file on the same disk drive, but for some reason it thinks it doesn't. A bug in the Go libraries, or what?!

**Comment by harshavardhana** [comment]
> Then I don't understand what the error message means. It's obviously trying to rename/move a file on the same disk drive, but for some reason it thinks it doesn't. A bug in the Go libraries, or what?!

@mhjartstrom We need to reproduce it - We have no idea why this happens. It should work automatically.

**Comment by mhjartstrom** [comment]
Well, if you have no idea how to reproduce it, then you'll have to tell me how I can help?

**Comment by harshavardhana** [comment]
Let's start with few details

- Windows version
- NTFS version

Can you reproduce this let's say with the previous cloud berry version?

**Comment by harshavardhana** [comment]
Also can you try creating a file with `$` sign on your NTFS volume at F: ?

**Comment by mhjartstrom** [comment]
* Windows 7 Pro SP1 (v6.1 b7601)
* NTFS v3.1
* Executing `echo Hello World > $` from the Command Prompt on `F:\` creates a file called `$` with proper contents.

As I said in the Regression section, my feeling is that it might be related to a late CloudBerry Backup release. The last successful backup was somewhere near 2019-08-12 and from what I can tell from the cbb.log file it seems I installed an update to v6.1.1.74 on 2019-08-10 and then the latest v6.1.3.9 on 2019-09-06. Neither of them would work, if I'm right.

Another fact that is pointing in that direction is that it did work fine a couple of months back, but now if I try downgrading to MinIO "2019-02-26T19:51:46Z" I still get the same error messages now.

**Comment by harshavardhana** [comment]
Have you spoken to CBB about this ? They have MinIO connector to natively support MinIO deployments.

**Comment by mhjartstrom** [comment]
Indeed I have. I started out with sending them the CBB detailed logs and creating a support issue, but when they found huge amounts of HTTP error code 500 (internal server error) they meant that it's a server side issue and asked me to investigate that instead.

**Comment by harshavardhana** [comment]
Did you say it worked before?  Can you downgrade to the known version of CBB and test again?

**Comment by mhjartstrom** [comment]
I'd have to verify that it's permissible to downgrade CBB without compromising the configuration. Also, I don't know where to find a download archive of the install executables. Same thing with MinIO actually, where can I find those?

**Comment by harshavardhana** [comment]
You can find them here https://dl.minio.io/server/minio/release/linux-amd64/archive/

**Comment by mhjartstrom** [comment]
No can do. I just went out on a limb and tried to downgrade from v6.1.3.9 to v6.0.2.22 (which was the only old version I had laying around) and it failed to interpret the configuration, quite obviously not compatible. I also read somewhere in a user forum that they do not provide download of old executables.

**Comment by harshavardhana** [comment]
That is unfortunate @mhjartstrom - will do a code review and see if anything broke. But only thing changed in this area is we moved to go1.13

**Comment by mhjartstrom** [comment]
It would be good if CloudBerry also got involved in this. I sent them a link to this issue.

**Comment by harshavardhana** [comment]
> That is unfortunate @mhjartstrom - will do a code review and see if anything broke. But only thing changed in this area is we moved to go1.13

Since you also said that this error is now reproducible also on older MinIO releases. looks like something has changed in CBB in the new release..

**Comment by mhjartstrom** [comment]
CloudBerry replied (regrettably privately instead of within this issue):

"We made some research on this, and it seems like other users are experiencing the same issue in this thread:  https://github.com/minio/minio/issues/4655

Additionally, it is referring to this one:  https://github.com/golang/go/issues/13766

It seems like this issue is related to GO packages on your current system which are causing this problem, as we haven't changed anything regarding MinIO storage in our last releases."

**Comment by mhjartstrom** [comment]
I don't know... There must be better ways to debug this issue?!

**Comment by harshavardhana** [comment]
To test further can you test this with any other Windows system?

**Comment by mhjartstrom** [comment]
I've come a lot further and have more or less found the "root cause" now (although not really).

There's a third player involved that I had not told you about and I'm sorry for that, but I simply did not think it had anything to do with these error messages. I use a utility called MirrorFolder (https://www.techsoftpl.com/backup/) for real-time (software) mirroring of specified folders in "RAID-1 type" mode of operation, and I have successfully done so for years.

I have configured MirrorFolder to mirror the contents of my MinIO object storage folder from one drive to another as well, e.g. from `F:\MinIO` to `G:\MirrorFolder\MinIO`.

It was only recently that I did a specific reconfiguration of the mirrors for the MinIO folders, in relation to my other open issue where MinIO wastes tremendous amounts of space and does not clean up its mess (#8129). I simply configured MirrorFolder to ignore mirroring the contents of the `tmp` folder of my MinIO object storage, since it's only a waste of space in the first place. This ought to be perfectly fine.

For some reason, this does not work out well when MinIO and the Go libraries try to do operations on the `tmp` folder, even though in my view MinIO shouldn't even be aware that there is a copy operation to another drive _being ignored_ at the time. Bare in mind that if I remove the exclude of the `tmp` folder in the object storage path, everything works just fine.

The simple solution is to remove the exclusion of the `tmp` directory from my MirrorFolder configuration, which again would waste _even more_ tremendous amounts of space for no apparent reason whatsoever. Hopefully you will close #8129 soon... But how come local MinIO operations on `F:` fail because the operation is _not_ being mirrored to `G:` simultaneously?

**Comment by harshavardhana** [comment]
> The simple solution is to remove the exclusion of the `tmp` directory from my MirrorFolder configuration, which again would waste _even more_ tremendous amounts of space for no apparent reason whatsoever. Hopefully you will close #8129 soon... But how come local MinIO operations on `F:` fail because the operation is _not_ being mirrored to `G:` simultaneously?

Because we hold exclusive locks on the operation on the backend @mhjartstrom and we need this for consistency requirements.

**Comment by mhjartstrom** [comment]
Somnath Kundu of Techsoft sent me a great explanation, although both he and I don't understand your last reply regarding exclusive locks:

"In RAID-1 type mirroring mode, MF duplicates file I/O (for writing only) at kernel level and send them to mirror device for duplication. This works fine for all file operations except file move/rename, where no data is copied. So, whenever you move a file inside a source folder, MF returns a status to Windows indicating that this is a different drive forcing Windows/application to copy the file so that we can mirror the file on the mirror drive. Note that the application that initiated move/rename of the file does not need to know about it as Windows ('MoveFile' 'MoveFileEx' APIs) will automatically handle this status transparently in the background. It seems MinIO is using 'MoveFileEx' API without 'MOVEFILE_COPY_ALLOWED' flag and hence this issue. But it is not clear to me how this has something to do with holding an exclude lock for this file move operation."

**Comment by harshavardhana** [comment]
@mhjartstrom we don't do anything special we just use os.Rename() and it is expected if the file changes underneath it should fail. We don't allow cross drive renames since those are not cross-platform friendly. 

So os.Rename() will work as long as tmp file and the destination file are both on the same drive. The problem here is that your backup software is holding a lock on the tmp file for some reason or adding some windows specific attributes by making it a LinkFile of sort. So the renames are failing. 

Backup software should be run through the S3 API, not through the backend. Because MinIO requires exclusive access to the backend if this is something you can't provide in your setup MinIO cannot be used safely.

**Comment by harshavardhana** [comment]
Regarding tmp files and they are leftover should be cleaned up properly. I will take a look at that issue right away.

**Comment by harshavardhana** [comment]
> Regarding tmp files and they are leftover should be cleaned up properly. I will take a look at that issue right away.

// cc @vadmeste ^^

**Comment by harshavardhana** [comment]
The issue is fixed and merged closing this as fixed in PR #8404
